raft/src/main/java/org/apache/kafka/raft/internals/EpochElection.java

```diff
@@ -0,0 +1,216 @@
+     */
+    public boolean isVoteRejected() {
+        return numGranted() + numUnrecorded() < majoritySize();
+    }
+
+    /**
+     * Get the set of voters which have not been counted as granted or rejected yet.
+     *
+     * @return The set of unrecorded voters
+     */
+    public Set<ReplicaKey> unrecordedVoters() {
+        return votersOfState(VoterState.State.UNRECORDED).collect(Collectors.toSet());
+    }
+
+    /**
+     * Get the set of voters that have granted our vote requests.
+     *
+     * @return The set of granting voters, which should always contain the localId
+     */
+    public Set<Integer> grantingVoters() {
+        return votersOfState(VoterState.State.GRANTED).map(ReplicaKey::id).collect(Collectors.toSet());
+    }
+
+    /**
+     * Get the set of voters that have rejected our candidacy.
+     *
+     * @return The set of rejecting voters
+     */
+    public Set<Integer> rejectingVoters() {
+        return votersOfState(VoterState.State.REJECTED).map(ReplicaKey::id).collect(Collectors.toSet());
+    }
+
+    private VoterState getVoterStateOrThrow(int voterId) {
+        VoterState voterState = voterStates.get(voterId);
+        if (voterState == null) {
+            throw new IllegalArgumentException("Attempt to access voter state of non-voter " + voterId);
+        }
+        return voterState;
+    }
+
+    private Stream<ReplicaKey> votersOfState(VoterState.State state) {
+        return voterStates
+            .values()
+            .stream()
+            .filter(voterState -> voterState.state().equals(state))
+            .map(VoterState::replicaKey);
+    }
+
+    private long numGranted() {
+        return votersOfState(VoterState.State.GRANTED).count();
+    }
+
+    private long numUnrecorded() {
+        return votersOfState(VoterState.State.UNRECORDED).count();
+    }
+
+    private int majoritySize() {
+        return voterStates.size() / 2 + 1;
+    }
+
+    @Override
+    public String toString() {
+        return String.format(
+            "EpochElection(%s)",
+            voterStates.values().stream()
+                .map(VoterState::toString)
+                .collect(
+                    Collectors.joining(", "))
+        );
+    }
+
+    private static final class VoterState {
+        private final ReplicaKey replicaKey;
+        private State state = State.UNRECORDED;
+
+        VoterState(ReplicaKey replicaKey) {
+            this.replicaKey = replicaKey;
+        }
+
+        public State state() {
+            return state;
+        }
+
+        public void setState(State state) {
+            this.state = state;
+        }
+
+        public ReplicaKey replicaKey() {
+            return replicaKey;
+        }
+
+        enum State {
+            UNRECORDED,
+            GRANTED,
+            REJECTED
+        }
+
+        @Override
+        public String toString() {
+            return String.format(
+                "VoterState(%s, state=%s)",
```

Comment:
```java
                "VoterState(replicaKey=%s, state=%s)",
```

Reply:
ReplicaKey's toString method contains the class name so I didn't want to be redundant - `String.format("ReplicaKey(id=%d, directoryId=%s)", id, directoryId);`